zipkin/src/main/java/zipkin/internal/Node.java

```diff
@@ -132,6 +132,8 @@ public void remove() {
     Map<Long, Long> idToParent = new LinkedHashMap<>(idToNode.size());
 
     public void addNode(@Nullable Long parentId, long id, V value) {
+      // drop circular deps vs NPE later
+      if (parentId != null && parentId.equals(id)) return;
```

Comment:
Since we're dropping datas, shouldn't we log that there's something wrong ? IMO, dropping data is a good way to deal with this problem but people need to know that their datas have a problem